### ISSUE TITLE
Obtaining whitened images without style transfer

### ISSUE THREAD

**Issue by engharat** [Yijunmaverick/UniversalStyleTransfer#6]
Hi,
I'm trying to get the whitened content images without applying any target style, as you have shown as example in the paper.
I'm working on the pytorch version, and to make things simple I have simply removed the code portion that I thought would apply the style transfer, returning whiten_CF instead of targetFeature:
`      ......
       whiten_cF = torch.mm(step2,cF)

        s_d = (s_e[0:k_s]).pow(0.5)
        targetFeature = torch.mm(torch.mm(torch.mm(s_v[:,0:k_s],torch.diag(s_d)),(s_v[:,0:k_s].t())),whiten_cF)
        targetFeature = targetFeature + s_mean.unsqueeze(1).expand_as(targetFeature)
        return targetFeature`
That should correspond in your original torch code at removing those lines:
`     
        local swap_latent = swap:forward(whiten_contentFeature:resize(sg[1], sg[2], sg[3])):clone()
        local swap_latent1 = swap_latent:view(sg[1], sg[2]*sg[3])
        targetFeature = (s_v[{{},{1,k_s}}]:cuda()) * (torch.diag(s_d1:cuda())) * (s_v[{{},{1,k_s}}]:t():cuda()) * swap_latent1`

Leaving in the function as last lines something like that:
`whiten_contentFeature = (c_v[{{},{1,k_c}}]:cuda()) * torch.diag(c_d:cuda()) * (c_v[{{},{1,k_c}}]:t():cuda()) *contentFeature1
    local Whiten_contentFeature= whiten_contentFeature:resize(sg[1], sg[2], sg[3])

    return Whiten_contentFeature
`
I tried this modification in the pytorch version and the output image is not whitened, but results the same as the content image. Could you suggest me how fix it?

**Comment by Yijunmaverick** [comment]
@engharat Thanks for your interests. It looks weird. I think your modification is right. Let me describe how I did that in my torch code.

In the _feature_wct_ function (not _feature_swap_whiten_) of the _test_wct.lua_ file, just return the _whiten_contentFeature_ (as you did):

```
    local tFeature = whiten_contentFeature:resize(sg[1], sg[2], sg[3])
    return tFeature
```
In the _styleTransfer_ function of the _test_wct.lua_ file (the whitened results in the paper are obtained by using the encoder4-decoder4 only):

```
    local cF4 = vgg4:forward(content):clone()
    local sF4 = vgg4:forward(style):clone()  (you can remove the style feature input accordingly)
    local csF4 = feature_wct(cF4, sF4)
    --csF4 = opt.alpha * csF4 + (1.0-opt.alpha) * cF4  (originally we have an "alpha" here to blend the csF4 with the content feature cF4, you can remove this line or set alpha=1)
    local Im4 = decoder4:forward(csF4)
    return Im4
```
I did a quick experiment and here is my test result:

Input:
![023](https://user-images.githubusercontent.com/10704714/36223170-62ca8278-1178-11e8-9264-911722836e6f.jpg)

Output:
![023_whiten](https://user-images.githubusercontent.com/10704714/36223184-759ad128-1178-11e8-8587-5dd6766b0856.jpg)

Hope you can fix it :)

**Comment by engharat** [comment]
Thanks for the very quick reply! Going down on the code another time, I catched a stupid mistake I made. Using all encoder-decoder from 5 to 1, I'm getting:
![pic1](https://user-images.githubusercontent.com/16682919/36223737-a5a75d80-11c5-11e8-9163-d1e8b5f0fefb.jpg)
![pic_whitened](https://user-images.githubusercontent.com/16682919/36223746-a91e8542-11c5-11e8-9c16-7dda2802f2d0.jpg)
That I suppose being the correct result. 
I'll try using only encoder/decoder 4,  to see if the whitened image keep more information about the original image - using encoder-decoder from 5 to 1 seems to keep very little features from original image.
